File: main/java/uk/co/real_logic/sbe/otf/OtfHeaderDecoder.java
/*
 * Copyright 2013 Real Logic Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package uk.co.real_logic.sbe.otf;

import uk.co.real_logic.sbe.PrimitiveType;
import uk.co.real_logic.sbe.codec.java.DirectBuffer;
import uk.co.real_logic.sbe.ir.HeaderStructure;
import uk.co.real_logic.sbe.ir.Token;

import java.nio.ByteOrder;

/**
 * Used to decode a message header while doing on-the-fly decoding of a message stream.
 * <p/>
 * Meta data is cached to improve the performance of decoding headers.
 */
public class OtfHeaderDecoder
{
    private final int size;
    private int templateIdOffset;
    private int templateVersionOffset;
    private int blockLengthOffset;
    private PrimitiveType templateIdType;
    private PrimitiveType templateVersionType;
    private PrimitiveType blockLengthType;
    private ByteOrder templateIdByteOrder;
    private ByteOrder templateVersionByteOrder;
    private ByteOrder blockLengthByteOrder;

    /**
     * Read the message header structure and cache the meta data for finding the key fields for decoding messages.
     *
     * @param headerStructure for the meta data describing the message header.
     */
    public OtfHeaderDecoder(final HeaderStructure headerStructure)
    {
        size = headerStructure.tokens().get(0).size();

        for (final Token token : headerStructure.tokens())
        {
            switch (token.name())
            {
                case HeaderStructure.TEMPLATE_ID:
                    templateIdOffset = token.offset();
                    templateIdType = token.encoding().primitiveType();
                    templateIdByteOrder = token.encoding().byteOrder();
                    break;

                case HeaderStructure.TEMPLATE_VERSION:
                    templateVersionOffset = token.offset();
                    templateVersionType = token.encoding().primitiveType();
                    templateVersionByteOrder = token.encoding().byteOrder();
                    break;

                case HeaderStructure.BLOCK_LENGTH:
                    blockLengthOffset = token.offset();
                    blockLengthType = token.encoding().primitiveType();
                    blockLengthByteOrder = token.encoding().byteOrder();
                    break;
            }
        }
    }

    /**
     * The size of the message header in bytes.
     *
     * @return the size of the message header in bytes.
     */
    public int size()
    {
        return size;
    }

    /**
     * Get the template id from the message header.
     *
     * @param buffer from which to read the value.
     * @param bufferOffset in the buffer at which the message header begins.
     * @return the value of the template id.
     */
    public int getTemplateId(final DirectBuffer buffer, final int bufferOffset)
    {
        return Util.getInt(buffer, bufferOffset + templateIdOffset, templateIdType, templateIdByteOrder);
    }

    /**
     * Get the template version number from the message header.
     *
     * @param buffer from which to read the value.
     * @param bufferOffset in the buffer at which the message header begins.
     * @return the value of the template version number.
     */
    public int getTemplateVersion(final DirectBuffer buffer, final int bufferOffset)
    {
        return Util.getInt(buffer, bufferOffset + templateVersionOffset, templateVersionType, templateVersionByteOrder);
    }

    /**
     * Get the block length of the root block in the message.
     *
     * @param buffer from which to read the value.
     * @param bufferOffset in the buffer at which the message header begins.
     * @return the length of the root block in the coming message.
     */
    public int getBlockLength(final DirectBuffer buffer, final int bufferOffset)
    {
        return Util.getInt(buffer, bufferOffset + blockLengthOffset, blockLengthType, blockLengthByteOrder);
    }
}

File: main/csharp/Otf/OtfHeaderDecoder.cs
using Adaptive.SimpleBinaryEncoding.ir;

namespace Adaptive.SimpleBinaryEncoding.Otf
{
    /// <summary>
    /// Used to decode a message header while doing on-the-fly decoding of a message stream.
    ///
    /// Meta data is cached to improve the performance of decoding headers.
    ///
    /// This class is thread safe.
    /// </summary>
    public class OtfHeaderDecoder
    {
        private readonly int _size;
        private readonly int _blockLengthOffset;
        private readonly int _templateIdOffset;
        private readonly int _schemaIdOffset;
        private readonly int _schemaVersionOffset;

        private readonly PrimitiveType _blockLengthType;
        private readonly PrimitiveType _templateIdType;
        private readonly PrimitiveType _schemaIdType;
        private readonly PrimitiveType _schemaVersionType;

        private readonly ByteOrder _blockLengthByteOrder;
        private readonly ByteOrder _templateIdByteOrder;
        private readonly ByteOrder _schemaIdByteOrder;
        private readonly ByteOrder _schemaVersionByteOrder;

        /// <summary>
        /// Read the message header structure and cache the meta data for finding the key fields for decoding messages.
        /// </summary>
        /// <param name="headerStructure"> for the meta data describing the message header. </param>
        public OtfHeaderDecoder(HeaderStructure headerStructure)
        {
            _size = headerStructure.Tokens[0].Size;

            foreach (Token token in headerStructure.Tokens)
            {
                switch (token.Name)
                {
                    case HeaderStructure.BlockLength:
                        _blockLengthOffset = token.Offset;
                        _blockLengthType = token.Encoding.PrimitiveType;
                        _blockLengthByteOrder = token.Encoding.ByteOrder;
                        break;

                    case HeaderStructure.TemplateId:
                        _templateIdOffset = token.Offset;
                        _templateIdType = token.Encoding.PrimitiveType;
                        _templateIdByteOrder = token.Encoding.ByteOrder;
                        break;

                    case HeaderStructure.SchemaId:
                        _schemaIdOffset = token.Offset;
                        _schemaIdType = token.Encoding.PrimitiveType;
                        _schemaIdByteOrder = token.Encoding.ByteOrder;
                        break;

                    case HeaderStructure.SchemaVersion:
                        _schemaVersionOffset = token.Offset;
                        _schemaVersionType = token.Encoding.PrimitiveType;
                        _schemaVersionByteOrder = token.Encoding.ByteOrder;
                        break;
                }
            }
        }

        /// <summary>
        /// The size of the message header in bytes.
        /// </summary>
        /// <value>the size of the message header in bytes.</value>
        public int Size
        {
            get { return _size; }
        }

        /// <summary>
        /// Get the template id from the message header.
        /// </summary>
        /// <param name="buffer"> from which to read the value. </param>
        /// <param name="bufferOffset"> in the buffer at which the message header begins. </param>
        /// <returns> the value of the template id. </returns>
        public int GetTemplateId(DirectBuffer buffer, int bufferOffset)
        {
            return Util.GetInt(buffer, bufferOffset + _templateIdOffset, _templateIdType, _templateIdByteOrder);
        }


        /// <summary>
        /// Get the schema id number from the message header.
        /// </summary>
        /// <param name="buffer">buffer from which to read the value.</param>
        /// <param name="bufferOffset">bufferOffset in the buffer at which the message header begins.</param>
        /// <returns>the value of the schema id number.</returns>
        public int GetSchemaId(DirectBuffer buffer, int bufferOffset)
        {
            return Util.GetInt(buffer, bufferOffset + _schemaIdOffset, _schemaIdType, _schemaIdByteOrder);
        }

        /// <summary>
        /// Get the schema version number from the message header.
        /// </summary>
        /// <param name="buffer"> from which to read the value. </param>
        /// <param name="bufferOffset"> in the buffer at which the message header begins. </param>
        /// <returns> the value of the schema version number. </returns>
        public virtual int GetSchemaVersion(DirectBuffer buffer, int bufferOffset)
        {
            return Util.GetInt(buffer, bufferOffset + _schemaVersionOffset, _schemaVersionType, _schemaVersionByteOrder);
        }

        /// <summary>
        /// Get the block length of the root block in the message.
        /// </summary>
        /// <param name="buffer"> from which to read the value. </param>
        /// <param name="bufferOffset"> in the buffer at which the message header begins. </param>
        /// <returns> the length of the root block in the coming message. </returns>
        public virtual int GetBlockLength(DirectBuffer buffer, int bufferOffset)
        {
            return Util.GetInt(buffer, bufferOffset + _blockLengthOffset, _blockLengthType, _blockLengthByteOrder);
        }
    }
}